daft/logical/schema2.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+
+from typing import Iterator, TypeVar
+
+from daft.daft import PyField as _PyField
+from daft.daft import PySchema as _PySchema
+from daft.datatype import DataType
+from daft.expressions2 import Expression, col
+
+ExpressionType = TypeVar("ExpressionType", bound=Expression)
+
+
+class Field:
+    _field: _PyField
+
+    def __init__(self) -> None:
+        raise NotImplementedError("We do not support creating a Field via __init__ ")
+
+    @staticmethod
+    def _from_pyfield(field: _PyField) -> Field:
+        f = Field.__new__(Field)
+        f._field = field
+        return f
+
+    @property
+    def name(self):
+        return self._field.name()
+
+    @property
+    def dtype(self) -> DataType:
+        return DataType._from_pydatatype(self._field.dtype())
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Field):
+            return False
+        return self._field.eq(other._field)
+
+
+class Schema:
+    _schema: _PySchema
+
+    def __init__(self) -> None:
+        raise NotImplementedError("We do not support creating a Schema via __init__ ")
+
+    @staticmethod
+    def _from_pyschema(schema: _PySchema) -> Schema:
+        s = Schema.__new__(Schema)
+        s._schema = schema
+        return s
+
+    def __getitem__(self, key: str) -> Field:
+        if key not in self._schema.names():
+            raise ValueError(f"{key} was not found in Schema of fields {self._schema.field_names()}")
+        pyfield = self._schema[key]
+        return Field._from_pyfield(pyfield)
+
+    def __len__(self) -> int:
+        return len(self._schema.names())
+
+    def column_names(self) -> list[str]:
+        return list(self._schema.names())
+
+    def __iter__(self) -> Iterator[Field]:
+        col_names = self.column_names()
+        return (self[name] for name in col_names)
```

Comment:
return `iter` of the tuple

Reply:
This actually returns a generator rather than a `tuple`:

```
>>> x = (i for i in range(10))
>>> type(x)
<class 'generator'>
```

I can do `yield from` instead if its clearer?

Reply:
nah its fine then